### ISSUE TITLE
Adding an fixed    Auxiliary Classifier loss   to  part the  lossD or lossG 

### ISSUE THREAD

**Issue by vgg4resnet** [junyanz/pytorch-CycleGAN-and-pix2pix#1084]
hi，
 I am trying to do an face recognition project 。Some of my Images are motion blured baddly，my trained face recognition model  mobilefacenet(this trained  model is fixed ) get bad results 。And I can not find paired images for pix2pix model，so I am using the cycle gan model to improve my images qualities。however it doesnt impove much。So  giving some image pairs from the same persons,the cyclegan model can generate some new imgs for those same persons pairs。And using the mobilefacnet recognition model， I  calculate the cos distance between images ，
  **and then adding  them together as  new loss  A 。I am trying to adding the loss A
 to loss D or loss G, I want decrease  the  as  new loss  A,the new part lossA drop slowly ，and can not converge ，the loss drop from 14  to 13 ，keep staying in around 13 **

**Comment by vgg4resnet** [comment]
this part code isgenerate news Images pairs,and  calucating the the loss A

`
    def infer_facepic(self):
        '''

        listSpeacialFiles=["3_0005.jpg","3_0007.jpg","3_0010.jpg","3_0012.jpg","3_0013.jpg",

        ]
        totalfaceloss=0
        iSum=0
        for i ,imgpys in enumerate( faces): 
               ####get imageA generate from cyclegan model 
                imgpy=imgpys[0]
                myimg=cv2.imread(imgpy[0])
                myimg=cv2.resize(myimg,(256,256))                
                    
                real_A=self.test_transform(myimg).to(self.device).unsqueeze(0)
                fake_B=self.test_transform(myimg).to(self.device).unsqueeze(0)
                pred_fake_B = tensor2im(self.netG_A(real_A))
                pred_rec_A = tensor2im(self.netG_B(fake_B))                             
                img=pred_fake_B
                ####get imageA's comparing face vectors
                embs1=self.facemodel(self.test_transform(img).to(self.devicess).unsqueeze(0)).detach().cpu().numpy().reshape(1,-1)


                imgpy=imgpys[1]
                myimg=cv2.imread(imgpy)

                myimg=Image.fromarray(myimg)
                real_A=self.test_transform(myimg).to(self.device).unsqueeze(0)
                fake_B=self.test_transform(myimg).to(self.device).unsqueeze(0)
                ####get imageB generate from cyclegan model 
                pred_fake_B = tensor2im(self.netG_A(real_A))
                pred_rec_A = tensor2im(self.netG_B(fake_B)) 
                img=pred_fake_B
                ####get imageB's comparing face vectors
                embs2=self.facemodel(self.test_transform(img).to(self.devicess).unsqueeze(0)).detach().cpu().numpy().reshape(1,-1)

 
                ####comparing A,B images loss
                totalfaceloss=totalfaceloss+1-cosine_similarity(embs1,embs2)[0][0]

        myReAutoGrad=self.lossReAutoGrad

        min_idx, minimum=0,0


        return min_idx, minimum ,myReAutoGrad  
`

**Comment by vgg4resnet** [comment]
inferring  every batch
![QQ截图20200701153426](https://user-images.githubusercontent.com/44825736/86216451-a7ad1d80-bbb0-11ea-907f-f55f8130d852.jpg)

**Comment by vgg4resnet** [comment]
`    def backward_D_basic(self, netD, real, fake):
        # Real
        pred_real = netD(real)
        loss_D_real = self.criterionGAN(pred_real, True)
        # Fake
        pred_fake = netD(fake.detach())
        loss_D_fake = self.criterionGAN(pred_fake, False)
        # Combined loss
        loss_D = (loss_D_real + loss_D_fake) * 0.5
        loss_D = (0*loss_D_real + 0*loss_D_fake+1*self.lossReAutoGrad) * 0.5
        # backward
        loss_D.backward()
        return loss_D`

**Comment by vgg4resnet** [comment]
I want to know how to decrea the self.lossReAutoGrad （the face recognition assuming name is  the loss A ），and I wonder adding the lossA to the lossD work。the loss can backward or not ?thanks in advanced

**Comment by junyanz** [comment]
Not sure if I  am completely following your implementation. But it seems that you need to use both GAN loss and face recognition loss. 
```
loss_D = (loss_D_real + loss_D_fake) * 0.5 + self. lossReAutoGrad
loss_D.backward()
```
Recognizing small faces is a challenging problem and out of my expertise. I recommend that you follow some popular methods such as this one (https://www.cs.cmu.edu/~peiyunh/tiny/)

**Comment by vgg4resnet** [comment]
yeah，thank your for your answer，loss_D = (loss_D_real + loss_D_fake) * 0.5 + self. lossReAutoGrad，I have tried that before，I  added ` (loss_D_real + loss_D_fake) * 0.5 + 7*self. lossReAutoGrad ` to make D loss more depend on face recognition loss。
 **the thing I want to know most is whether changing  D loss by this way,the loss can backward the network or not ,the gradient will be affected or not**

**Comment by junyanz** [comment]
I think there are two issues in your code. 
- You probably should add lossReAutoGrad to the `loss_G`, as only the generator is involved in the computational graph  of this loss. 
-   You should feed `pred_fake_B` directly to  `self.facemodel` (after applying some proper transformation in PyTorch).  Currently, `tensor2im` and `test_transform` will stop the gradients.

**Comment by vgg4resnet** [comment]
hi，according to what your say  I try the following code but still not works
```
self.infer_facepic()
self.fake_B = self.netG_A(self.real_A)
self.rec_A = self.netG_B(self.fake_B)
self.fake_A = self.netG_B(self.real_B)
self.rec_B = self.netG_A(self.fake_A)
```
by this way ，the gradients may not  be affected

`
    def backward_G(self):
        lambda_idt = self.opt.lambda_identity
        lambda_A = self.opt.lambda_A
        lambda_B = self.opt.lambda_B
        # Identity loss
        if lambda_idt > 0:
            # G_A should be identity if real_B is fed.
            self.idt_A = self.netG_A(self.real_B)
            self.loss_idt_A = self.criterionIdt(self.idt_A, self.real_B) * lambda_B * lambda_idt
            # G_B should be identity if real_A is fed.
            self.idt_B = self.netG_B(self.real_A)
            self.loss_idt_B = self.criterionIdt(self.idt_B, self.real_A) * lambda_A * lambda_idt
        else:
            self.loss_idt_A = 0
            self.loss_idt_B = 0

        # GAN loss D_A(G_A(A))
        fake_xishu=-1.0*-1
        if(1):
            
            self.loss_G_A = self.criterionGAN(self.netD_A(self.fake_B), True)+7*self.lossReAutoGrad  
            print("7*self.lossReAutoGrad ,self.loss_G_A",self.loss_G_A,7*self.lossReAutoGrad   )
            # GAN loss D_B(G_B(B))
            self.loss_G_B = self.criterionGAN(self.netD_B(self.fake_A), True)+0*self.lossReAutoGrad  
        else:
            self.loss_G_A = fake_xishu*self.criterionGAN(self.netD_A(self.fake_B), True)-7*self.lossReAutoGrad  
            print("7*self.lossReAutoGrad ,self.loss_G_A",self.loss_G_A,7*self.lossReAutoGrad   )
            # GAN loss D_B(G_B(B))
            self.loss_G_B = fake_xishu*self.criterionGAN(self.netD_B(self.fake_A), True)+0*self.lossReAutoGrad                  

        # Forward cycle loss
        self.loss_cycle_A = self.criterionCycle(self.rec_A, self.real_A) * lambda_A
        # Backward cycle loss
        self.loss_cycle_B = self.criterionCycle(self.rec_B, self.real_B) * lambda_B
        # combined loss
        #self.loss_G = self.loss_G_A + self.loss_G_B + self.loss_cycle_A + self.loss_cycle_B + self.loss_idt_A + self.loss_idt_B
        self.loss_G = self.loss_G_A + self.loss_G_B + self.loss_cycle_A + self.loss_cycle_B + self.loss_idt_A + self.loss_idt_B     
        self.loss_G.backward()`
may be pred_fake_B clone  tensor  copy_ function  work

**Comment by vgg4resnet** [comment]
any more help,by follow junyanz  suggestion ,my way seem not work，and  feed pred_fake_B directly to self.facemodel ，but image size to self.facemodel need is 112*112,our network is 256*256, how to do ,I am stucked

**Comment by junyanz** [comment]
You can resize the image using Pytorch function, before feeding to the model.

**Comment by vgg4resnet** [comment]
OK,i will try it ,thank you so much for your help

**Comment by junyanz** [comment]
maybe this one: https://pytorch.org/docs/stable/nn.functional.html#upsample
